src/sentry/monitors/clock_tasks/check_missed.py

```diff
@@ -0,0 +1,157 @@
+from sentry.monitors.logic.mark_failed import mark_failed
+from sentry.monitors.models import (
+    CheckInStatus,
+    MonitorCheckIn,
+    MonitorEnvironment,
+    MonitorStatus,
+    MonitorType,
+)
+from sentry.monitors.schedule import get_prev_schedule
+from sentry.utils import metrics
+
+from .producer import MONITORS_CLOCK_TASKS_CODEC, produce_task
+
+logger = logging.getLogger(__name__)
+
+
+# This is the MAXIMUM number of MONITOR this job will check.
+#
+# NOTE: We should keep an eye on this as we have more and more usage of
+# monitors the larger the number of checkins to check will exist.
+MONITOR_LIMIT = 10_000
+
+
+def dispatch_check_missing(ts: datetime):
+    """
+    Given a clock tick timestamp determine which monitor environments are past
+    their next_checkin_latest, indicating they haven't checked-in when they
+    should have
+
+    This will dispatch MarkMissing messages into monitors-clock-tasks.
+    """
+    qs = (
+        # Monitors that have reached the latest checkin time
+        MonitorEnvironment.objects.filter(
+            monitor__type__in=[MonitorType.CRON_JOB],
+            next_checkin_latest__lte=ts,
+        )
+        .exclude(
+            status__in=[
+                MonitorStatus.DISABLED,
+                MonitorStatus.PENDING_DELETION,
+                MonitorStatus.DELETION_IN_PROGRESS,
+            ]
+        )
+        .exclude(
+            monitor__status__in=[
+                ObjectStatus.DISABLED,
+                ObjectStatus.PENDING_DELETION,
+                ObjectStatus.DELETION_IN_PROGRESS,
+            ],
+        )
+        .exclude(
+            monitor__is_muted=True,  # Temporary fix until we can move out of celery or reduce load
+        )
+        .exclude(
+            is_muted=True,  # Temporary fix until we can move out of celery or reduce load
+        )[:MONITOR_LIMIT]
+    )
+
+    metrics.gauge("sentry.monitors.tasks.check_missing.count", qs.count(), sample_rate=1.0)
+    for monitor_environment in qs:
+        message: MarkMissing = {
+            "type": "mark_missing",
+            "ts": ts.timestamp(),
+            "monitor_environment_id": monitor_environment.id,
+        }
+        # XXX(epurkhiser): Partitioning by monitor_environment.id is important
+        # here as these task messages will be consumed in a multi-consumer
+        # setup. If we backlogged clock-ticks we may produce multiple missed
+        # tasks for the same monitor_environment. These MUST happen in-order.
+        payload = KafkaPayload(
+            monitor_environment.id.to_bytes(),
+            MONITORS_CLOCK_TASKS_CODEC.encode(message),
+            [],
+        )
+        produce_task(payload)
+
+
+def mark_environment_missing(monitor_environment_id: int, ts: datetime):
+    logger.info("mark_missing", extra={"monitor_environment_id": monitor_environment_id})
+
+    try:
+        monitor_environment = MonitorEnvironment.objects.select_related("monitor").get(
```

Comment:
Should we revalidate status and is muted etc?

Reply:
👍